# Pipeline basic

Here we will go through each step from the raw signals file (slow5) to generating plots.

For this we will use `r9.4_450bps.nucleotide` DNA dataset. The dataset is at `test/data/raw/pipelines/pipeline_0/r9.4_450bps.nucleotide`.
Note that the script must be executed inside the virtual environment where squigualiser is installed.

## Pipeline variables
The following variables must be changed before running the bash script.

| **Variable name**               | **Description**                                                                                                                                                                                                      | **value set for this pipeline**      |
|---------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------|
| SQUIGULATOR                     | path to squigulator                                                                                                                                                                                                  | -                                    |
| SAMTOOLS                        | path to samtools                                                                                                                                                                                                     | -                                    |
| MINIMAP2                        | path to minimap2                                                                                                                                                                                                     | -                                    |
| GUPPY                           | path to dir containing guppy executables                                                                                                                                                                             | -                                    |
| BUTTERY_EEL_ENV_PATH            | path to python env where buttery-eel is installed                                                                                                                                                                    | -                                    |
| REFERENCE                       | path to the reference genome file                                                                                                                                                                                    | -                                    |
| MODEL_TO_USE                    | this has to be set to the name of the model used to basecall                                                                                                                                                         | dna_r9.4.1_450bps_fast_prom.cfg      |
| CHUNK_SIZE                      | (optional) This is used in basecalling to limit GPU memory.                                                                                                                                                          | 500                                  |
| SIMULATING_PROFILE              | signal profile used by squigulator (-x argument)                                                                                                                                                                     | dna-r9-prom                          |
| RE_REFORM_K                     | `kmer_length` value used in the second run of `reform`. For new data this value can only be set after running `calculate_offsets` step (see pipeline steps) More details can be found at [profiles](profiles.md)     | 3                                    |
| RE_REFORM_M                     | `sig_move_offset` value used in the second run of `reform`. For new data this value can only be set after running `calculate_offsets` step (see pipeline steps) More details can be found at [profiles](profiles.md) | 2                                    |
| PROFILE_TO_DETERMINE_BASE_SHIFT | `squigualiser` uses this value to determine the appropriate `base_shift`. More details can be found at [profiles](profiles.md)                                                                                       | kmer_model_dna_r9.4.1_450bps_6_mer   |
| SIGNAL_FILE                     | SLOW5/BLOW5 signal file                                                                                                                                                                                              | reads.blow5                          |
| READ_ID                         | An arbitrary read_id from the SIGNAL_FILE to perform `calculate_offsets_read_id` step (see pipeline steps)                                                                                                           | 8e1a33c4-af69-471c-a115-6428c8bf63df |
| READ_REGION                     | The region to simulate and plot in the selected read.                                                                                                                                                                | ${READ_ID}:1-500                     |
| REF_REGION                      | The region in the reference to simulate and plot. This region cannot be determined without running `minimap2` step. User can load the minimap2 aligned sam file to IGV and select a region.                          | chr16:46,389,459-46,390,388          |
| SIM_REGION                      | (optional) The region to plot in simulated read and the reference.                                                                                                                                                   | sim_ref:1-500                        |
| SIG_SCALE                       | (optional) `squigualiser` scales the signal. More information at [commands](commands.md)                                                                                                                             | znorm                                |

## Pipeline steps
Each step listed in the bash script `run.sh` is discussed below.

| **Step**                        | **Description**                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
|---------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| basecalling                     | `Buttery-eel` is used to basecall and create moves.sam. Sam file is then converted to bam and fastq for later steps.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| reform_move_table               | run `reform` with `-k 1` and `-m 0`.  Optionally, to skip steps 4, 5, and 6 provide the correct --profile argument. For this pipeline lets not skip the steps.                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| calculate_offsets               | use `reform` output to determine best k and m values.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| calculate_offsets_read_id       | Cretae a pdf file with the distributions to visualise the separation (useful for the user to further confirm if the determined k and m values are optimal).                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| re_reform                       | Run `reform` again using the determined k and m values.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| simulate_read_signal            | Simulate a read using `squigulator`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| plot_sim_and_real_signal        | Plot both simulated and real signals in two separate tracks.  Two pileup commands (one for plotting simulated and one for real signal) were added to a file.  Note `--profile` argument is provided for the simulated signal plot. The profile name depends on what kmer model was used to simulate the signal. For the real signal plot `--profile` argument is not necessary as we have ran reform again with new k and m values to make the base shift 0. Then the file is given as the input to plot_tracks command.                                                                                            |
| minimap2_align                  | Align the basecalled reads to the reference.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| realign                         | Run `realign` tool to re-align cigar mapping to moves.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
| simulate_ref_signal             | simulate part of the reference region using `squigulator`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| plot_realign_and_sim            | Plot both simulated reference and real signals mapped to the reference (using realign) in two separate tracks.  Two pileup commands (one for plotting simulated and one for real signals) were added to a file.  Note `--profile` argument is provided for the simulated signal plot. The profile name depends on what kmer model was used to simulate the signal. For the real signal plot `--profile` argument is not necessary as we have ran reform again with new k and m values to make the base shift 0. Then the file is given as the input to plot_tracks command.                                         |
| f5c_eventalign                  | Peform `f5c eventalign` to get the eventalignment file.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| plot_eventalign_and_sim         | Plot both simulated reference and real signals mapped to the reference in two separate tracks.  Two pileup commands (one for plotting simulated and one for real signals) were added to a file.  Note `--profile` argument is provided for the simulated signal plot. The profile name depends on what kmer model was used to simulate the signal. In this case, also for the real signal plot `--profile` argument is necessary as we `f5c` used a kmer model.  Since the same kmer model was used in `squigulator` and `f5c`, profile name the same.  Then the file is given as the input to plot_tracks command. |
| plot_eventalign_forward_reverse | So far all the plots only used forward mapped reads as it is the default. However, in this step forward and reverse mapped reads are plottted using the f5c eventalignment output. Note that `--plot_reverse` argument is provided in the second command. As usual, profile name is also provided.                                                                                                                                                                                                                                                                                                                  |
| plot_eventalign_forward_reverse | Similar to step `plot_eventalign_forward_reverse` but this time using `realign` data instead of `f5c` eventalignment. Hence, profile name is not required.                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

You can use this basic pipeline script as a start and do the necessary changes to support your analysis.

A similar pipeline is available for `dna_r10.4.1_e8.2_400bps` data at `test/data/raw/pipelines/pipeline_0/dna_r10.4.1_e8.2_400bps`.
A similar pipeline is available for `rna_r9.4.1_70bps` data at `test/data/raw/pipelines/pipeline_0/rna_r9.4.1_70bps`.
